File: examples/port_scanner/port_scanner.py
#!/usr/bin/env python

import os, resource
import asyncio

class PortScanner:
    def __init__(self, host="0.0.0.0", ports=range(1, 1024+1), batch_size=1024):
        self.host       = host
        self.ports      = ports
        self.semaphore  = asyncio.Semaphore(value=batch_size)
        self.loop       = asyncio.get_event_loop()

    async def scan_port(self, port, timeout):
        async with self.semaphore:
            try:
                future = asyncio.open_connection(self.host, port, loop=self.loop)
                reader, writer = await asyncio.wait_for(future, timeout=timeout)
                print("{} open".format(port))
                writer.close()
            except ConnectionRefusedError:
                print("{} closed".format(port))
            except asyncio.TimeoutError:
                print("{} timeout".format(port))

    def start(self, timeout=1.0):
        self.loop.run_until_complete(asyncio.gather(
            *[self.scan_port(port, timeout) for port in self.ports]
            ))

limits = resource.getrlimit(resource.RLIMIT_NOFILE)
batch_size = min(512, limits[0])

scanner = PortScanner(ports=range(1, 65536), batch_size=batch_size)

scanner.start()

File: examples/port_scanner/port_scanner.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Released under the MIT License.
# Copyright, 2018, by Kent 'picat' Gruber.
# Copyright, 2018-2025, by Samuel Williams.

require "async/io"
require "async/semaphore"
require_relative "../../lib/async/await"

class PortScanner
	include Async::Await
	include Async::IO

	def initialize(host: "0.0.0.0", ports:, batch_size: 1024)
		@host      = host
		@ports     = ports
		@semaphore = Async::Semaphore.new(batch_size)
	end

	def scan_port(port, timeout)
		with_timeout(timeout) do
			address = Async::IO::Address.tcp(@host, port)
			peer = Socket.connect(address)
			puts "#{port} open"
			peer.close
		end
		rescue Errno::ECONNREFUSED
		# puts "#{port} closed"
		rescue Async::TimeoutError
			puts "#{port} timeout"
	end

	async def start(timeout = 1.0)
		@ports.map do |port|
			@semaphore.async do
				scan_port(port, timeout)
			end
		end.collect(&:result)
	end
end

limits = Process.getrlimit(Process::RLIMIT_NOFILE)
batch_size = [512, (limits.first * 0.9).ceil].min

scanner = PortScanner.new(host: "127.0.0.1", ports: Range.new(1, 65535), batch_size: batch_size)

scanner.start
